### ISSUE TITLE
Fix "sudo: command not found"

### ISSUE THREAD

**Issue by iainelder** [iainelder/dotfiles#5]
[Today's PowerShell CI run failed](https://github.com/iainelder/dotfiles/actions/runs/8595037344/job/23549056909).

The presenting error is:

`programs/powershell/install.bash: line 10: sudo: command not found`

The root cause appears to be in the "Prepare Docker" step.

```
+ apt-get --yes update
Get:1 http://security.ubuntu.com/ubuntu focal-security InRelease [114 kB]
Get:2 http://archive.ubuntu.com/ubuntu focal InRelease [265 kB]
Get:3 http://archive.ubuntu.com/ubuntu focal-updates InRelease [114 kB]
Get:4 http://archive.ubuntu.com/ubuntu focal-backports InRelease [108 kB]
Err:1 http://security.ubuntu.com/ubuntu focal-security InRelease
  Unknown error executing apt-key
Err:2 http://archive.ubuntu.com/ubuntu focal InRelease
  Unknown error executing apt-key
Err:3 http://archive.ubuntu.com/ubuntu focal-updates InRelease
  Unknown error executing apt-key
Err:4 http://archive.ubuntu.com/ubuntu focal-backports InRelease
  Unknown error executing apt-key
Reading package lists...
W: GPG error: http://security.ubuntu.com/ubuntu focal-security InRelease: Unknown error executing apt-key
E: The repository 'http://security.ubuntu.com/ubuntu focal-security InRelease' is not signed.
W: GPG error: http://archive.ubuntu.com/ubuntu focal InRelease: Unknown error executing apt-key
E: The repository 'http://archive.ubuntu.com/ubuntu focal InRelease' is not signed.
W: GPG error: http://archive.ubuntu.com/ubuntu focal-updates InRelease: Unknown error executing apt-key
E: The repository 'http://archive.ubuntu.com/ubuntu focal-updates InRelease' is not signed.
W: GPG error: http://archive.ubuntu.com/ubuntu focal-backports InRelease: Unknown error executing apt-key
E: The repository 'http://archive.ubuntu.com/ubuntu focal-backports InRelease' is not signed.
```

The output comes from this compound command.

```bash
apt-get --yes update && apt-get --yes install sudo
```

The first part `apt-get --yes update` updates the local copy of the repository index.

The second part `apt-get --yes install sudo` installs sudo.

Because the first part fails, the second part never starts.

I thought the line `set -euxo pipefail` would configure Bash to halt if the first part failed like that, but apparently not.

---

`apt-get  --yes update` writes four failures with this pattern:

```
Err:1 ...
  Unknown error executing apt-key
W: GPG error: ... InRelease: Unknown error executing apt-key
E: The repository '...' is not signed.
```

For four Canonical repos:

```
http://security.ubuntu.com/ubuntu focal-security InRelease
http://archive.ubuntu.com/ubuntu focal InRelease
http://archive.ubuntu.com/ubuntu focal-updates InRelease
http://archive.ubuntu.com/ubuntu focal-backports InRelease
```

Complete example for first repo:

```
Get:1 http://security.ubuntu.com/ubuntu focal-security InRelease [114 kB]
Err:1 http://security.ubuntu.com/ubuntu focal-security InRelease
  Unknown error executing apt-key
Reading package lists...
W: GPG error: http://security.ubuntu.com/ubuntu focal-security InRelease: Unknown error executing apt-key
E: The repository 'http://security.ubuntu.com/ubuntu focal-security InRelease' is not signed.
```

**Comment by iainelder** [comment]
I saw the same error in [clustergit CI run 576 attempt 1](https://github.com/iainelder/dotfiles/actions/runs/8586474066/job/23528983040).

**Comment by iainelder** [comment]
Similar error in my local environment.

The `apt-get --yes update` appears to fail, but not hard enough to skip the execution of `apt-get --yes install sudo`.

This time it's `apt-get --yes install sudo` that fails.

```console
$ ./test_local.bash
+ script=_
++ pwd
+ docker run -it --rm --volume /var/run/docker.sock:/var/run/docker.sock --mount type=bind,source=/home/isme/Repos/dotfiles,target=/code ubuntu:20.04 bash -c '/code/scripts/install_in_docker.bash _'
+ program=_
+++ dirname /code/scripts/install_in_docker.bash
++ cd /code/scripts
++ pwd
+ script_dir=/code/scripts
+ /code/scripts/prepare_docker.bash
+ apt-get --yes update
Err:1 http://security.ubuntu.com/ubuntu focal-security InRelease         
  Temporary failure resolving 'security.ubuntu.com'
Err:2 http://archive.ubuntu.com/ubuntu focal InRelease                   
  Temporary failure resolving 'archive.ubuntu.com'
Err:3 http://archive.ubuntu.com/ubuntu focal-updates InRelease
  Temporary failure resolving 'archive.ubuntu.com'
Err:4 http://archive.ubuntu.com/ubuntu focal-backports InRelease
  Temporary failure resolving 'archive.ubuntu.com'
Reading package lists... Done        
W: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/focal/InRelease  Temporary failure resolving 'archive.ubuntu.com'
W: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/focal-updates/InRelease  Temporary failure resolving 'archive.ubuntu.com'
W: Failed to fetch http://archive.ubuntu.com/ubuntu/dists/focal-backports/InRelease  Temporary failure resolving 'archive.ubuntu.com'
W: Failed to fetch http://security.ubuntu.com/ubuntu/dists/focal-security/InRelease  Temporary failure resolving 'security.ubuntu.com'
W: Some index files failed to download. They have been ignored, or old ones used instead.
+ apt-get --yes install sudo
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Unable to locate package sudo
```

**Comment by iainelder** [comment]
The presenting error happens because `set -e` doesn't just exit as soon as a command returns a non-zero status.

[The Bash manual](https://www.gnu.org/software/bash/manual/bash.html#The-Set-Builtin) lists many caveats. Among them, the shell does not exit if the command that fails is part of any command executed in a `&&` or `||` list except the command following the final `&&` or `||`.

An AND list has the form `command1 && command2`. command2 is executed if, and only if, command1 returns an exit status of zero (success).

An OR list has the form `command1 || command2`. command2 is executed if, and only if, command1 returns a non-zero exit status. 

Use this script to try all possible combinations.

```bash
function generate-cases() {
    printf "%s\n" {false,true}" && "{false,true}
    printf "%s\n" {false,true}" || "{false,true}
}

function testcase() {
    output=$(bash -c "set -e; $1; echo \$?")
    printf "%s: %s\n" "$1" "$output"
}

while read cmd; do
    testcase "$cmd"
done < <(generate-cases)
```

```
false && false: 1
false && true: 1
true && false: 
true && true: 0
false || false: 
false || true: 0
true || false: 0
true || true: 0
```

> Because the first part fails, the second part never starts.

That's the short-circuiting in action as described in the manual.

> I thought the line `set -euxo pipefail` would configure Bash to halt if the first part failed like that

No, it's one of `-e` documented caveats.

> The `apt-get --yes update` appears to fail, but not hard enough to skip the execution of `apt-get --yes install sudo`.

The command's status is unclear. The output shows "Err" (error) and "W" (warning?) messages. Perhaps a warning is still a 0 status. I'd need to read the apt manual and find a reliable repro.

---

It seems the solution here is to avoid `&&` and just write each command on a new line. In this way the script will halt when the any simple command fails.

**Comment by iainelder** [comment]
Probably fixed by [7d61fb8](https://github.com/iainelder/dotfiles/commit/7d61fb85ab672441dfe5d4e203acc02bc6bc285b) and maybe by #7.
